File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/WeavingFilter.java
package com.codepoetics.raffia.streaming;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.writers.BasketWeavingWriter;
import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

public abstract class WeavingFilter<T extends BasketWriter<T>> implements FilteringWriter<T> {

  private int depth = 0;
  private final FilteringWriter<T> parent;
  private BasketWeavingWriter weaver;

  protected WeavingFilter(FilteringWriter<T> parent, BasketWeavingWriter weaver) {
    this.parent = parent;
    this.weaver = weaver;
  }

  @Override
  public FilteringWriter<T> key(String key) {
    weaver = weaver.key(key);
    return this;
  }

  @Override
  public FilteringWriter<T> add(String value) {
    weaver = weaver.add(value);
    return this;
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    weaver = weaver.add(value);
    return this;
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    weaver = weaver.add(value);
    return this;
  }

  @Override
  public FilteringWriter<T> addNull() {
    weaver = weaver.addNull();
    return this;
  }

  @Override
  public T complete() {
    throw new IllegalStateException("Cannot complete() while still weaving basket");
  }

  @Override
  public FilteringWriter<T> beginObject() {
    depth += 1;
    weaver = weaver.beginObject();
    return this;
  }

    @Override
    public FilteringWriter<T> beginArray() {
      depth += 1;
      weaver = weaver.beginObject();
      return this;
    }

    @Override
    public FilteringWriter<T> end() {
      weaver = weaver.end();

      return depth-- == 0 ? parent.advance(writeToTarget(weaver.weave())) : this;
    }

    protected abstract T writeToTarget(Basket woven);

  @Override
  public FilteringWriter<T> advance(T newTarget) {
    throw new IllegalStateException("Cannot advance() while weaving basket");
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/WeavingFilter.kt
package com.codepoetics.raffia.streaming

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.writers.BasketWeavingWriter
import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

abstract class WeavingFilter<T : BasketWriter<T>> protected constructor(private val parent: FilteringWriter<T>, private var weaver: BasketWeavingWriter?) : FilteringWriter<T> {

    private var depth = 0

    override fun key(key: String): FilteringWriter<T> {
        weaver = weaver!!.key(key)
        return this
    }

    override fun add(value: String): FilteringWriter<T> {
        weaver = weaver!!.add(value)
        return this
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        weaver = weaver!!.add(value)
        return this
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        weaver = weaver!!.add(value)
        return this
    }

    override fun addNull(): FilteringWriter<T> {
        weaver = weaver!!.addNull()
        return this
    }

    override fun complete(): T {
        throw IllegalStateException("Cannot complete() while still weaving basket")
    }

    override fun beginObject(): FilteringWriter<T> {
        depth += 1
        weaver = weaver!!.beginObject()
        return this
    }

    override fun beginArray(): FilteringWriter<T> {
        depth += 1
        weaver = weaver!!.beginObject()
        return this
    }

    override fun end(): FilteringWriter<T> {
        weaver = weaver!!.end()

        return if (depth-- == 0) parent.advance(writeToTarget(weaver!!.weave())) else this
    }

    protected abstract fun writeToTarget(woven: Basket): T

    override fun advance(newTarget: T): FilteringWriter<T> {
        throw IllegalStateException("Cannot advance() while weaving basket")
    }
}

